impl/src/main/java/org/ehcache/config/builders/CacheConfigurationBuilder.java

```diff
@@ -169,17 +170,61 @@ private CacheConfigurationBuilder(CacheConfigurationBuilder<K, V> other) {
    * @return a new builder with the added service configuration
    *
    * @see #add(ServiceConfiguration)
+   * @deprecated in favor of {@link #withService(Builder)}
    */
-  public CacheConfigurationBuilder<K, V> add(Builder<? extends ServiceConfiguration<?>> configurationBuilder) {
+  @Deprecated
+  public CacheConfigurationBuilder<K, V> add(Builder<? extends ServiceConfiguration<?, ?>> configurationBuilder) {
     return add(configurationBuilder.build());
   }
 
-  /**
-   * Adds an {@link EvictionAdvisor} to the returned builder.
-   *
-   * @param evictionAdvisor the eviction advisor to be used
-   * @return a new builder with the added eviction advisor
-   */
+  @Override
+  public <C extends ServiceConfiguration<?, ?>> Collection<C> getServices(Class<C> configurationType) throws IllegalArgumentException {
+    return serviceConfigurations.stream().filter(configurationType::isInstance).map(configurationType::cast).collect(toList());
+  }
+
+  @Override
+  public CacheConfigurationBuilder<K, V> withService(ServiceConfiguration<?, ?> config) {
+    CacheConfigurationBuilder<K, V> otherBuilder = new CacheConfigurationBuilder<>(this);
+    otherBuilder.serviceConfigurations.removeIf(other -> !other.compatibleWith(config) || !config.compatibleWith(other));
+    otherBuilder.serviceConfigurations.add(config);
+    return otherBuilder;
+  }
+
+  @Override
+  public CacheConfigurationBuilder<K, V> withoutServices(Class<? extends ServiceConfiguration<?, ?>> clazz) {
+    CacheConfigurationBuilder<K, V> otherBuilder = new CacheConfigurationBuilder<>(this);
+    otherBuilder.serviceConfigurations.removeIf(clazz::isInstance);
+    return otherBuilder;
+  }
+
+  @Override
+  public <C extends ServiceConfiguration<?, ?>>CacheConfigurationBuilder<K, V> withoutServices(Class<C> clazz, Predicate<? super C> predicate) {
+    CacheConfigurationBuilder<K, V> otherBuilder = new CacheConfigurationBuilder<>(this);
+    otherBuilder.serviceConfigurations.removeIf(c -> clazz.isInstance(c) && predicate.test(clazz.cast(c)));
+    return otherBuilder;
+  }
+
+  @Override
+  public <R, C extends ServiceConfiguration<?, R>> CacheConfigurationBuilder<K, V> updateServices(Class<C> clazz, UnaryOperator<R> update) {
+    Collection<? extends ServiceConfiguration<?, R>> existing = getServices(clazz);
+
+    if (existing.isEmpty()) {
+      throw new IllegalStateException("Cannot update service configurations. No services exist");
```

Comment:
Should the service that isn't found be identified?

Reply:
Makes sense. Done.